### ISSUE TITLE
Import function and autoload PHP source file

### ISSUE THREAD

**Issue by ParkFramework** [marcioAlmada/RFCs#5]
Hi
Proposal syntax to autoload source file to imported functions.
If without slash end namespace, use this namespace as autoload file.
Like this:

``` php
use function foo\math {
 sin, cos, cosh 
};

sin($int); // __autoload('foo\math')
```

Maybe other option, to be able to call other functions that were not explicitly mentioned in the declaration use.
Like this:

``` php
use foo\math;

math\sqrt($int);  // __autoload('foo\math')

```

Combined option:

``` php
// the "math" imported to current namespace, if without slash end.
use function foo\math {
 sin, cos, cosh 
};

// __autoload('foo\math')
sin($int);

// this function is loaded, in \foo\math namespace
math\sqrt($int);
```

What do you think about this?

**Comment by marcioAlmada** [comment]
Hi,

> Proposal syntax to autoload source file to imported functions.
> If without slash end namespace, use this namespace as autoload file.
> 
> ``` php
> use foo\math;
> math\sqrt($int);  // __autoload('foo\math')
> ```

@ircmaxell is already planning an RFC regarding function autoloading based on the previous withdrawn one: https://wiki.php.net/rfc/function_autoloading.

> Maybe other option, to be able to call other functions that were not explicitly mentioned in the declaration use.
> Like this:
> 
> ``` php
> use foo\math;
> math\sqrt($int);  // __autoload('foo\math')
> ```

This is already possible, see: http://3v4l.org/X7Itb

So your combined example can be achieved with this:

``` php
use function foo\math\{sin, cos, cosh};
use foo\math;

sin($int);
math\sqrt($int);
```

I don't see any improvement on your suggestions considering what we already have at hands.

The only serious improvement I can think of is to allow `use foo\math\*` and then let the user land autoloader implementation handle the `*` wildcard somehow, perhaps I should give it a try.

Thoughts?

**Comment by ParkFramework** [comment]
> This is already possible, see: http://3v4l.org/X7Itb

Yes, but it does not cause the autoloader:
http://3v4l.org/nN7f2

It is necessary to cause the autoloader, as it happens when you call the class method:
http://3v4l.org/016SE

> The only serious improvement I can think of is to allow use foo\math* and then let the user land autoloader implementation handle the \* wildcard somehow, perhaps I should give it a try.

Yes, looks good, is similar in other languages.

But I want to just be able to not write slash at the end of use namespace.

**Comment by marcioAlmada** [comment]
> Yes, but it does not cause the autoloader:
> http://3v4l.org/nN7f2
> It is necessary to cause the autoloader, as it happens when you call the class method:
> http://3v4l.org/016SE

Yes, sure, I understand you but this is **completely** orthogonal to group use.

You have a good reason to contact Anthony and ask about his plans regarding `__autoload` and his next functional autoloading RFC https://wiki.php.net/rfc/function_autoloading, this is a problem that should be resolved independently from syntax.

> Yes, looks good, is similar in other languages.
> But I want to just be able to not write slash at the end of use namespace.

I won't be the one proposing the removal of the slash for aesthetic reasons. I've been importing functional apis in real world code and the slash is ok.

``` php
use combinator\{Token, TokenStream, Directives};

use function combinator\parsers\{
    either, chain, any, token, repeat, optional, between, operator, backtrack, commit
    braces, brackets, parentheses
};
```

PS: the syntax highlighters tend to paint the opening `{` with the wrong color, that's the annoying part, but tools will be updated ;)

**Comment by marcioAlmada** [comment]
TL;DR

If your problem is functional **autoloading**, contact the author of the functional autoloading RFC. Not long ago, he told me he needed help to finish the new proposal and the patch.

But if your issue is basically "drop the backslash", sorry, I disagree, this is not worth a language change (community already voted for it)  + people should have picked a better namespace separator years ago :smile:

**Comment by ParkFramework** [comment]
Ok :)
